### PR TITLE
Support transform dialect matchers in dispatch region formation

### PR DESCRIPTION
**Pull request by ftynse**
This requires some changes to the base logic of the transform dialect interpreter pass. The transform IR is constructed on the fly in a separate module, the ownership of which is shared by all copies of the dispatch-with-interpreter pass, similarly to the externally supplied transform IR file. The interpreter currently is currently run multiple times with Linalg ops as potential roots to control the matching from C++ instead of deferring it to the transform dialect.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:43]: Please use:
```
std::pair<Value, Value> mlir::iree_compiler::buildSelectFirstNonEmpty(
    ImplicitLocOpBuilder &b, Value handle1, Value handle2) {
```

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:49]: All this could be a 1-liner if MovePrecedingOpIntoDispatchRegionOp was extended to allow unpacking the `getRest` handle on the fly (did that for some other op).

If you don't want to do it here, could you add to the TODO section of the running doc ?

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h:153]: Why not just `return SmallVector<Operation*>` ?

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (nicolasvasilache) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:104]: Nit: pass loc explicitly and lift up the comment 1 level, no more need to pass context.

**Review by nicolasvasilache — COMMENTED** [review]
You probably want in the same commit to go into:
```
compiler/src/iree/compiler/Dialect/Flow/Transforms/Passes.cpp
```

and change the logic around : 
```

      .addPredicatedPass(!clDispatchTransformFileName.empty(),
                         [&]() {
                           return createDispatchWithTransformDialect(
                               clDispatchTransformFileName);
                         })
```

to add a flag and perform dispatch formation in the absence of a file.

Then you'll want some extra test(s) that run the first steps of the traditional iree pipeline with this flag on to make sure that the transform-based dispatch formation composes properly with falling back to the default IREE one in cases we don't handle.

You prob. want both changes in the same commit so we avoid annoying rebases as we iterate in the branch.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h:167]: Some comment here to explain use cases ?

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h:181]: This comment prob needs some more explanation, I am not able to link parsing or construction to the rest of the documentation in this file at first glance.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp:1229]: maybe add a sentence about slicing IR from a larger graph into a dispatch region vs performing codegen of exactly all the things previously sliced into a dispatch region: we reuse the same code for both.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/IR/StructuredTransformOpsExt.cpp:1277]: Can we have a basic wrapping mechanism that just forwards and adds true? 
This way we don't need to declare a linear number of these trivial indirection functions in the future.

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreterPassBase.cpp
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreterPassBase.cpp:327]: frankly .. just `signalPassFailure` .. I fear that with in some form or another this message will be drowned in other noise and we'll waste cycles figuring things out until a forehead slap moment

File: llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreterPassBase.cpp
Comment (nicolasvasilache) [inline llvm-external-projects/iree-dialects/lib/Dialect/LinalgTransform/Passes/TransformInterpreterPassBase.cpp:18]: I'd just drop all this and only allow debug in single threaded mode.
In the future all this will need to be more tightly integrated with IREE anyway and hopefully transform IR attached separately and not dropped on the floor by the nested pass pipeline.

**Review by nicolasvasilache — APPROVED** [review]
Nice, feel free to land once addressed.

Note to all: this is landing in the transform-dialect-bringup branch

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (ftynse) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:49]: Not exactly one-liner, but changed this to merge handles + single move.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h
Comment (ftynse) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h:153]: This is the preferred way in LLVM/MLIR and the code here is intended to go there.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (ftynse) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:43]: This would make Flow depend on Codegen, I suppose this is not desirable from the layering perspective.

Comment (ftynse) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp:43]: I'll move that to iree-dialects.

File: llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h
Comment (ftynse) [reply llvm-external-projects/iree-dialects/include/iree-dialects/Dialect/LinalgTransform/TransformInterpreterPassBase.h:181]: Added more information above, here is the wrong place.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp
Comment (nicolasvasilache) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/DispatchWithTransformDialect.cpp]: ah .. I saw rest and thought tail .. 

Would be great to have some nicer API to hide all this mess but fine as a followup.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before `LinalgFoldUnitExtentDimsPass`. This should be a proceporcessing step.

**Comment by MaheshRavishankar** [comment]
> It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before `LinalgFoldUnitExtentDimsPass`. This should be a proceporcessing step.



> It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before `LinalgFoldUnitExtentDimsPass`. This should be a proceporcessing step.

Saw Nicolas' comment. THanks! Ok unblocking, but some clarity before this lands in main would be useful....

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by ftynse** [comment]
> It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before LinalgFoldUnitExtentDimsPass. This should be a proceporcessing step.

We want a possibility to override the formation of dispatch regions that codegen knows how to handle. The specific example Nicolas has is the 5/6-op softmax.

I haven't looked at the whole pipeline as the pass stub was already there before my changes, but I agree that it would make more sense to run this earlier in the pipeline, at least before FusionOfTensorOps. Not sure about LinalgFoldUnitExtentDimsPass. Depending on the patterns used in custom formation, it may or may not be desirable to do that as "canonicalization": e.g., do we want to "stength-reduce" a matmul with unit dimension so it's matched as matvec or not. Any thoughts?

> Saw Nicolas' comment. THanks! Ok unblocking, but some clarity before this lands in main would be useful....

I am happy to keep discussing after sending this to the branch to unblock Nicolas.

Specifically, I'd like your input (and potentially @benvanik) on layering. I don't like the hack I had to do in the build system. We have Flow transforms that include all transformation logic and pass definitions. Flow transform ops need to depend on transform logic, but not pass definitions. They do because there's only one library. Transform op application pass needs to depend on Flow transform ops, which would create a cycle. IMO, a better layering is to split transformation logic from pass definitions, but it is rather intrusive. Suggestions?

**Comment by stellaraccident** [comment]
> > It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before LinalgFoldUnitExtentDimsPass. This should be a proceporcessing step.
> 
> We want a possibility to override the formation of dispatch regions that codegen knows how to handle. The specific example Nicolas has is the 5/6-op softmax.
> 
> I haven't looked at the whole pipeline as the pass stub was already there before my changes, but I agree that it would make more sense to run this earlier in the pipeline, at least before FusionOfTensorOps. Not sure about LinalgFoldUnitExtentDimsPass. Depending on the patterns used in custom formation, it may or may not be desirable to do that as "canonicalization": e.g., do we want to "stength-reduce" a matmul with unit dimension so it's matched as matvec or not. Any thoughts?
> 
> > Saw Nicolas' comment. THanks! Ok unblocking, but some clarity before this lands in main would be useful....
> 
> I am happy to keep discussing after sending this to the branch to unblock Nicolas.
> 
> Specifically, I'd like your input (and potentially @benvanik) on layering. I don't like the hack I had to do in the build system. We have Flow transforms that include all transformation logic and pass definitions. Flow transform ops need to depend on transform logic, but not pass definitions. They do because there's only one library. Transform op application pass needs to depend on Flow transform ops, which would create a cycle. IMO, a better layering is to split transformation logic from pass definitions, but it is rather intrusive. Suggestions?

There is indeed a layering issue here. I don't have an answer right now but that is because I'm producing y thinking about some other things, not because I don't think it needs attention.

If we can avoid committing to anything load bearing that will be hard to undo here until I can get some more time to digest it, that would be ideal.

**Comment by MaheshRavishankar** [comment]
> > It might not be directly related to the change itself, but I need a little bit more clarification on the goal with this effort..... I understand this is an escape hatch and this is needed in some cases, but I would rather move this all the way before LinalgFoldUnitExtentDimsPass. This should be a proceporcessing step.
> 
> We want a possibility to override the formation of dispatch regions that codegen knows how to handle. The specific example Nicolas has is the 5/6-op softmax.

On this point, I think a lot of collective mind share has gone into specifics of softmax fusion. On a different PR the solution from Harsh (from Nod) where an explicit softmax op is added, which implements the TilingInterface, and expands the op into its constituent ops is being explored. That seems to be one way of handling custom fusions. We might not need explicit ops for each custom fusion. Matchers might provide a way of matching "pre-cooked" DAGs that correspond to known good patterns. Like any other tool, we should use the tool for the right purpose. If we end up writing hundreds of matchers, for things that could be encompassed by a simpler logic (which is what the default fusion heuristics in IREE tries to do by trying to reduce dependence distance using elementwise op fusion, or tile + fuse when there are common parallel loops), that would not be a good end-state IMO... But there is definitely a place for such custom fusions on particular backends.

**Comment by ftynse** [comment]
> On this point, I think a lot of collective mind share has gone into specifics of softmax fusion. On a different PR the solution from Harsh (from Nod) where an explicit softmax op is added, which implements the TilingInterface, and expands the op into its constituent ops is being explored. That seems to be one way of handling custom fusions. We might not need explicit ops for each custom fusion.

We discussed exactly this with Stella a couple of days ago. This goes back to usual question of abstraction/dialect design and what is reasonable to have as ops, how many are acceptable, etc. Softmax is a "natural" thing to encode. Some other fusions may be purely a compiler artifact. It would be interesting to have some sort of transient ops that we don't need to define.

> Matchers might provide a way of matching "pre-cooked" DAGs that correspond to known good patterns. Like any other tool, we should use the tool for the right purpose. If we end up writing hundreds of matchers, for things that could be encompassed by a simpler logic (which is what the default fusion heuristics in IREE tries to do by trying to reduce dependence distance using elementwise op fusion, or tile + fuse when there are common parallel loops), that would not be a good end-state IMO... But there is definitely a place for such custom fusions on particular backends.

I definitely agree that an end state entirely driven by case-specific matchers isn't desirable. However, I expect that customized compilation logic per subgraph and/or per backend will be added no matter what. So it is preferable to have this customization visible, clearly specify the cases it handles, and be easy to disable, rather than being diluted in the general pipeline that may thus end up much more complex than the simple distance-minimization logic. At the very least, it will be clear when there is a hundred patterns as opposed to seven innocent branches in the flow, which combinatorially result in the same number of choices.

**Comment by MaheshRavishankar** [comment]
> > On this point, I think a lot of collective mind share has gone into specifics of softmax fusion. On a different PR the solution from Harsh (from Nod) where an explicit softmax op is added, which implements the TilingInterface, and expands the op into its constituent ops is being explored. That seems to be one way of handling custom fusions. We might not need explicit ops for each custom fusion.
> 
> We discussed exactly this with Stella a couple of days ago. This goes back to usual question of abstraction/dialect design and what is reasonable to have as ops, how many are acceptable, etc. Softmax is a "natural" thing to encode. Some other fusions may be purely a compiler artifact. It would be interesting to have some sort of transient ops that we don't need to define.
> 
> > Matchers might provide a way of matching "pre-cooked" DAGs that correspond to known good patterns. Like any other tool, we should use the tool for the right purpose. If we end up writing hundreds of matchers, for things that could be encompassed by a simpler logic (which is what the default fusion heuristics in IREE tries to do by trying to reduce dependence distance using elementwise op fusion, or tile + fuse when there are common parallel loops), that would not be a good end-state IMO... But there is definitely a place for such custom fusions on particular backends.
> 
> I definitely agree that an end state entirely driven by case-specific matchers isn't desirable. However, I expect that customized compilation logic per subgraph and/or per backend will be added no matter what. So it is preferable to have this customization visible, clearly specify the cases it handles, and be easy to disable, rather than being diluted in the general pipeline that may thus end up much more complex than the simple distance-minimization logic. At the very least, it will be clear when there is a hundred patterns as opposed to seven innocent branches in the flow, which combinatorially result in the same number of choices.

I overall agree. But I see this more of an escape hatch, than something that should become core to the compiler development. I'd tread cautiously here. There is a conscious effort made in IREE that fusion decisions are made using general characteristics of Linalg ops, with the explicit goal to incrementally increase the complexity of the DAG that is given to the backend as a unit of compilation. Within the backend we might need some "loose" matchers to chose between 10s of stratergies, but all with similar properties. I realize that what these are havent been documented, but its only recently that things have stabilized there (and the whole softmax discussion, just muddled the picture a lot....). I do plan to explain this further in a shared setting.

**Pull request closed without merging**
